File: src/harkerrobolib/subsystems/HSShooter.java
package harkerrobolib.subsystems;

import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.StatorCurrentLimitConfiguration;
import com.ctre.phoenix.motorcontrol.can.TalonFX;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import harkerrobolib.util.Conversions;
import harkerrobolib.util.Gains;
import harkerrobolib.util.Conversions.SpeedUnit;

/**
 * Represents the Shooter Subsystem, controlled by two Falcon 500s.
 * 
 * @author Jatin Kohli
 * @author Shahzeb Lakhani 
 * @author Anirudh Kotamraju
 * @author Arjun Dixit
 * @author Aimee Wang
 * @author Rohan Bhowmik
 * @author Chirag Kaushik
 * 
 * @since April 25, 2020
 */
public abstract class HSShooter extends SubsystemBase {
    private TalonFX master;
    private TalonFX follower;

    private boolean sensorPhase;
    private int gearRatio;
    private double wheelDiameter;
    private int ticksPerRev;

    /**
     * Constructs a shooter
     * 
     * @param master The master Falcon
     * @param follower The follower Falcon
     * @param sensorPhase The sensor phase for the master Falcon encoder
     * @param gearRatio The gear ratio from motor to flywheel
     * @param wheelDiameter The diameter of the shooter wheels
     * @param ticksPerRev The amount of encoder ticks per revolution
     */
    public HSShooter(TalonFX master, TalonFX follower,
        boolean sensorPhase, int gearRatio, double wheelDiameter, int ticksPerRev) {
        this.master = master;
        this.follower = follower;
        this.sensorPhase = sensorPhase;

        this.gearRatio = gearRatio;
        this.wheelDiameter = wheelDiameter;
        this.ticksPerRev = ticksPerRev;
        
        setupFlywheel();
    }

    /**
     * Sets the master and follower inverts
     * @param masterInverted the master invert
     * @param followerInverted the follower invert
     */
    public void setInverted(boolean masterInverted, boolean followerInverted) {
        master.setInverted(masterInverted);
        follower.setInverted(followerInverted);
    }
    
    /**
     * Configs the voltage comp
     * @param voltageComp to be set
     * @param enable whether or not to enable voltage compensation
     */
    public void configVoltageComp(double voltageComp, boolean enable) {
        master.configVoltageCompSaturation(voltageComp);
        master.enableVoltageCompensation(enable);
    }

    /**
     * Configs stator current limits
     * @param continuousLimit The continuous current limit
     * @param peakLimit The peak current limit
     * @param peakDur The amount of time that the motor is allowed 
     * to peak before output is stopped
     */
    public void configStatorCurrentLimit(double continuousLimit, double peakLimit, double peakDur) {
        master.configStatorCurrentLimit(
            new StatorCurrentLimitConfiguration(true, continuousLimit, peakLimit, peakDur));
    }

    /**
     * Sets up the master and follower talons.
     */
    public void setupFlywheel() {
        master.configFactoryDefault();
        follower.configFactoryDefault();
        
        follower.follow(master);
        
        master.setNeutralMode(NeutralMode.Coast);
      
        master.setSensorPhase(sensorPhase);

        master.configForwardSoftLimitEnable(false);
        master.configReverseSoftLimitEnable(false);
        master.overrideLimitSwitchesEnable(false);
    }

    /**
     * Set up for Velocity PID.
     * 
     * @param velocitySlot the slot for flywheel velocity
     * @param pidGains the pid constants to be set
     */
    public void setupVelocityPID(int velocitySlot, Gains pidGains) {
        master.config_kF(velocitySlot, pidGains.getkF());
        master.config_kP(velocitySlot, pidGains.getkP());
        master.config_kI(velocitySlot, pidGains.getkI());
        master.config_kD(velocitySlot, pidGains.getkD());
        master.config_IntegralZone(velocitySlot, pidGains.getIZone());
    }

    /**
     * Spins the shooter flywheel at a certain percent output
     * 
     * @param percentOutput The percent output to set the motors to
     */
    public void spinShooterPercentOutput(double percentOutput) {
        if(percentOutput == 0) 
            master.set(ControlMode.Disabled, 0);
        else
            master.set(ControlMode.PercentOutput, percentOutput);
    }

    /**
     * Spins the shooter flywheel at a certain velocity (in feet/second)
     * 
     * @param velocity the velocity to set
     */
    public void spinShooterVelocity(double velocity) {
        double velocityInTicks = Conversions.convertSpeed(SpeedUnit.FEET_PER_SECOND, velocity * gearRatio, SpeedUnit.ENCODER_UNITS, wheelDiameter, ticksPerRev);
        if(0.95 * velocityInTicks > master.getSelectedSensorVelocity())
            master.set(ControlMode.PercentOutput, 1);
        else 
            master.set(ControlMode.Velocity, velocityInTicks);
    }

    /**
     * Gets the master falcon.
     * 
     * @return the master falcon
     */
    public TalonFX getMaster() {
        return master;
    }

    /**
     * Gets the follower falcon.
     * 
     * @return the follower falcon
     */
    public TalonFX getFollower() {
        return follower;
    }
}

File: src/harkerrobolib/subsystems/HSShooter.kt
package harkerrobolib.subsystems

import com.ctre.phoenix.motorcontrol.ControlMode
import com.ctre.phoenix.motorcontrol.FeedbackDevice
import com.ctre.phoenix.motorcontrol.IMotorController
import com.ctre.phoenix.motorcontrol.NeutralMode

import edu.wpi.first.wpilibj.command.Subsystem
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard
import harkerrobolib.util.Gains
import harkerrobolib.wrappers.HSPigeon
import harkerrobolib.wrappers.HSTalon

/**
 * Represents a Shooter subsystem for the offseason shooter bot
 *
 * @author Finn Frankis
 * @author Jatin Kohli
 *
 * @param leftTalon the left Talon.
 * @param rightTalon the right Talon.
 *
 * @version 11/18/18
 */
abstract class HSDrivetrain
(
        val leftTalon: HSTalon,
        val rightTalon: HSTalon
        ) : Subsystem() {

    /**
     * Represents direction of intake.
     *
     * @param signum sign for output of left intake with given direction (must be negated for right)
     */
    enum class IntakeDirection(val signum : Int) {
        IN(1), OUT(-1);
        operator fun times(other : Double) = signum * other
    }
	
    /**
     * Applies a given operation to both Talons (see lambdas in Java for more clarification).
     * @param consumer the operation to be applied.
     */
    fun applyToTalons(lambda : (HSTalon) -> Unit) {
        lambda.invoke(leftTalon)
        lambda.invoke(rightTalon)
    }
	
    /**
     * Sets all active controllers to a given neutral mode (brake or coast).
     * @param nm the neutral mode to be applied.
     */
    fun setNeutralMode(nm: NeutralMode) {
        applyToTalons({ controller : HSTalon -> controller.setNeutralMode(nm) })

    }

    /**
     * Configures both Talons for current limiting.
     * @param peakLimit the peak limit (temporary, to account for current spikes).
     * @param peakTime the time for which the peak limit is active
     * @param continuousLimit the continuous limit, or the one which is active after the conclusion of the peak limit.
     * @param timeout the time after which a failed current limit command will stop being retried.
     */
    @JvmOverloads
    fun setCurrentLimit(peakLimit: Int, peakTime: Int, continuousLimit: Int, timeout: Int = -1) {
        val currentLimit = { talon : HSTalon ->
            val newTimeout = if (timeout == -1) talon.defaultTimeout else timeout
            talon.configPeakCurrentLimit(peakLimit, newTimeout)
            talon.configPeakCurrentDuration(peakTime, newTimeout)
            talon.configContinuousCurrentLimit(continuousLimit, newTimeout)
            Unit
        }
        applyToTalons(currentLimit)
    }

    /**
     * Resets both talons.
     */
    fun resetTalons() {
        applyToTalons({ talon -> talon.reset() })
    }


    /**
     * Configures the sensor for the Talons to point to a given sensor.
     * @param fd the type of sensor to which the Talon should use
     * @param pidLoop the loop index where this sensor should be placed [0,1]
     */
    fun configFeedbackSensor(fd: FeedbackDevice, pidLoop: Int) {
        leftTalon.configSelectedFeedbackSensor(fd,
                pidLoop)
        rightTalon.configSelectedFeedbackSensor(fd,
                pidLoop)
    }

    /**
     * Prints the current output percentage to the motors to SmartDashboard.
     */
    fun printMotorOutputPercentage() {
        SmartDashboard.putNumber("Left Talon Output Percentage", leftTalon.motorOutputPercent)
        SmartDashboard.putNumber("Right Talon Output Percentage", rightTalon.motorOutputPercent)
    }

    /**
     * Prints the closed loop error of the Talons in a given loop.
     * @param pidLoop the loop index [0,1]
     */
    fun printClosedLoopError(pidLoop: Int) {
        SmartDashboard.putNumber("Left Talon Closed Loop Error " + if (pidLoop == 0) "Primary" else "Auxiliary", leftTalon.getClosedLoopError(pidLoop).toDouble())
        SmartDashboard.putNumber("Right Talon Closed Loop Error " + if (pidLoop == 0) "Primary" else "Auxiliary", rightTalon.getClosedLoopError(pidLoop).toDouble())
    }

    /**
     * Prints the sensor positions of the Talons in a given loop.
     * @param pidLoop the loop index [0,1]
     */
    fun printSensorPositions(pidLoop: Int) {
        SmartDashboard.putNumber("Left Talon Position " + if (pidLoop == 0) "Primary" else "Auxiliary", leftTalon.getSelectedSensorPosition(pidLoop).toDouble())
        SmartDashboard.putNumber("Right Talon Position" + if (pidLoop == 0) "Primary" else "Auxiliary", rightTalon.getSelectedSensorPosition(pidLoop).toDouble())
    }

    /**
     * Determines whether the closed loop error for both sides is within a given value.
     * @param loopIndex the loop index, either primary or auxiliary [0,1]
     * @param allowableError the error tolerance to be checked
     * @return true if the absolute value of the error is within the value; false otherwise
     */
    fun isClosedLoopErrorWithin(loopIndex: Int, allowableError: Double): Boolean {
        return Math.abs(leftTalon.getClosedLoopError(loopIndex)) < allowableError && Math.abs(rightTalon.getClosedLoopError(loopIndex)) < allowableError
    }

    /**
     * Configures closed loop constants for the two Talons.
     * @param slotIndex the PID slot index [0,3] where the constants will be stored
     * @param leftConstants the set of constants for the left Talon
     * @param rightConstants the set of constants for the right Talon
     */
    fun configClosedLoopConstants(slotIndex: Int, leftConstants: Gains, rightConstants: Gains) {
        arrayOf("kF", "kP", "kI", "kD", "iZone").forEach {
            leftTalon.javaClass.getMethod("config_${it}", Int.javaClass, Double.javaClass).invoke(leftTalon, slotIndex,
                    leftConstants.javaClass.getField(it).get(leftTalon) as Double)
            rightTalon.javaClass.getMethod("config_${it}", Int.javaClass, Double.javaClass).invoke(rightTalon, slotIndex,
                    rightConstants.javaClass.getField(it).get(rightTalon) as Double)
        }
    }

    /**
     * Sets wheels to the same output and direction.
     *
     * @param output value from controller that sets output for left and right wheels to same value
     * @param direction sets both wheels to either IN or OUT direction
     */
    fun intakeOuttake(output: Double, direction: IntakeDirection) {
        if (direction == IntakeDirection.IN) {
            intakeOuttakeCube(output, output, IntakeDirection.IN, IntakeDirection.IN)
        } else {
            intakeOuttakeCube(output, output, IntakeDirection.OUT, IntakeDirection.OUT)
        }

    }

    /**
     * Sets each wheel to its specified output.
     *
     * @param leftOutput value that sets output for the left wheel
     * @param rightOutput value that sets output for the right wheel
     */
    fun intakeOuttake(leftOutput: Double, rightOutput: Double) {
        leftTalon.set(ControlMode.PercentOutput, leftOutput)
        rightTalon.set(ControlMode.PercentOutput, rightOutput)
    }

    /**
     * Sets wheels to their respective outputs and directions.
     *
     * @param leftOutput value from controller that sets left wheel output
     * @param rightOutput value from controller that sets right wheel output
     * @param leftDirection enum representing left wheel's direction
     * @param rightDirection enum representing right wheel's direction
     */
    fun intakeOuttake(leftOutput: Double, rightOutput: Double, leftDirection: IntakeDirection, rightDirection: IntakeDirection) {
        if (leftDirection == IntakeDirection.IN) {
            leftTalon[ControlMode.PercentOutput] = leftDirection * leftOutput
        } else {
            leftTalon[ControlMode.PercentOutput] = leftDirection * -leftOutput
        }
        if (rightDirection == IntakeDirection.IN) {
            rightTalon[ControlMode.PercentOutput] = rightDirection * -rightOutput
        } else {
            rightTalon[ControlMode.PercentOutput] = rightDirection * rightOutput
        }

    }
}